### PR TITLE
dsg-1664/new-schema-page

### PR DESCRIPTION
**Pull request by SouradeepSaha**
# Basics

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Styling and layout changes
- [ ] Code style changes (formatting, local variables)
- [ ] Refactoring (no functional changes)
- [ ] Build or CI related changes
- [ ] Tests
- [ ] Other... Please describe:

## What changes does this PR introduce?
- Created a page that enables creating a new schema. 
- Ensures uniqueness of keys by returning error if user tries to create a schema with a key that already exists
- Schema can be only created if key does not exist
- Added test cases for appropriate components
- Used Form errors to check for errors

## Screenshot of the affected changes

Screenshot 2023-10-11 at 10 23 06 AM

Screenshot 2023-10-11 at 11 03 15 AM

## Checklist

- [x] I have updated the documentation accordingly
- [x] I have confirmed all acceptance criteria defined by the story have been met
- [ ] I have confirmed that Sonar Qube findings for new code are reviewed and fixed, or will be whitelisted

### CODE REVIEW

**Review by derekdiep42 — COMMENTED** [review]
Check out my [PR](https://github.com/Nuvalence/dsgov-web/pull/487), theres a lot of subtle differences in how you're doing it that isn't consistent with the merged stuff. IE: validation, how you're initiating the formGroup (validation/length) and other stuff in the TS/HTML file. Focus on the TS/HTML files.

File: libs/agency/feature/admin/src/lib/components/schemas-list/schemas-list.component.ts [hunk 1]
```diff
@@ -147,6 +147,10 @@ export class SchemasListComponent {
     }
   }
 
+  public navigateToAddSchema(): void {
+    this._router.navigate(['/admin/schemas/add-schema']);
```
Comment (derekdiep42) [inline libs/agency/feature/admin/src/lib/components/schemas-list/schemas-list.component.ts:151]:
```suggestion
    this._router.navigate(['/admin', 'schemas', 'create']);
```

File: libs/agency/feature/admin/src/lib/components/schemas-list/schema-form/schema-form.component.html [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+    <header class="schema-form-header">
+      <h1 class="schema-form-header-title">Create Schema</h1>
+    </header>
+
+    <div class="schema-form-fields" [formGroup]="schemaFormGroup">
+      <nuverial-text-input
+        class="schema-form-field"
+        [required]="true"
+        [validationMessages]="{required: 'Key is required'}"
```
[the remark this answers is not in the document]
Comment (SouradeepSaha) [reply libs/agency/feature/admin/src/lib/components/schemas-list/schema-form/schema-form.component.html:17]: done

File: libs/agency/feature/admin/src/lib/components/schemas-list/schema-form/schema-form.component.ts [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+    schemaModel.name = formValue.name ?? '';
+
+    this._workApiRoutesService
+      .getSchemaDefinitionByKey$(schemaModel.key)
+      .pipe(
+        take(1),
+        tap(_ => {
+          this._nuverialSnackBarService.notifyApplicationError();
+        }),
+        catchError(_error => {
```
[the remark this answers is not in the document]
Comment (SouradeepSaha) [reply libs/agency/feature/admin/src/lib/components/schemas-list/schema-form/schema-form.component.ts:133]: Resolved based on conversation

File: libs/agency/feature/admin/src/lib/components/schemas-list/schema-form/schema-form.component.spec.ts [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+        {
+          provide: Router,
+          useValue: {
+            navigate: jest.fn(),
+          },
+        },
+        {
+          provide: LoggingAdapter,
+          useValue: mockLoggingService,
+        },
```
Comment (Ortega-Dan) [inline libs/agency/feature/admin/src/lib/components/schemas-list/schema-form/schema-form.component.spec.ts:59]: not sure if this is being used. But may not matter, just bringing it up, but can be just "nit".

Comment (SouradeepSaha) [reply libs/agency/feature/admin/src/lib/components/schemas-list/schema-form/schema-form.component.spec.ts:59]: I get this error if I remove these:

Screenshot 2023-10-12 at 1 14 28 PM

**Comment by Ortega-Dan** [comment]
Thanks @SouradeepSaha . It looks good to me at this point :+1:  Leaving it for final review by @davidyoung77 or @odeljanin

File: libs/agency/feature/admin/src/lib/components/schemas-list/schema-form/schema-form.component.ts [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+    NuverialTextAreaComponent,
+    NuverialFormErrorsComponent,
+    ReactiveFormsModule,
+  ],
+  selector: 'dsg-schema-form',
+  standalone: true,
+  styleUrls: ['./schema-form.component.scss'],
+  templateUrl: './schema-form.component.html',
+})
+@UntilDestroy()
```
Comment (davidyoung77) [inline libs/agency/feature/admin/src/lib/components/schemas-list/schema-form/schema-form.component.ts:61]: You aren't using this, please remove

File: libs/agency/feature/admin/src/lib/components/schemas-list/schema-form/schema-form.component.ts [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+
+  public createSchema() {
+    this.formErrors = [];
+    if (!this.schemaFormGroup.valid) {
+      this.formErrors = FormErrorsFromGroup(this.schemaFormGroup, this.formConfigs);
+      MarkAllControlsAsTouched(this.schemaFormGroup);
+
+      return;
+    }
+
```
Comment (davidyoung77) [inline libs/agency/feature/admin/src/lib/components/schemas-list/schema-form/schema-form.component.ts:135]: We need to make sure we are handling loading at this point

**Review by davidyoung77 — COMMENTED** [review]
I have a couple comments then this will be good

File: libs/agency/feature/admin/src/lib/components/schemas-list/schema-form/schema-form.component.ts (hunk 6 above)
Comment (SouradeepSaha) [reply libs/agency/feature/admin/src/lib/components/schemas-list/schema-form/schema-form.component.ts:135]: Sorry but could you please clarify what we mean by loading? I'm not really sure I understand, is it the spinner?

Comment (davidyoung77) [reply libs/agency/feature/admin/src/lib/components/schemas-list/schema-form/schema-form.component.ts:135]: To clarify are you handling the the loading state/spinner while saving?

**Review by davidyoung77 — APPROVED** [review]

**Pull request merged**